src/serialization/preprocessor.cpp

```diff
@@ -1124,26 +1165,77 @@ bool preprocessor_data::get_chunk()
 			}
 			else if (target_.depth_ < 100 && (macro = target_.defines_->find(symbol)) != target_.defines_->end())
 			{
-				preproc_define const &val = macro->second;
+				const preproc_define &val = macro->second;
 				size_t nb_arg = strings_.size() - token.stack_pos - 1;
-				if (nb_arg != val.arguments.size())
+				size_t optional_arg_num = 0;
+
+				std::map<std::string, std::string> *defines = new std::map<std::string, std::string>;
+				const std::string& dir = filesystem::directory_name(val.location.substr(0, val.location.find(' ')));
+
+				for (size_t i = 0; i < nb_arg; ++i) {
+					if (i < val.arguments.size()) {
+						// Normal mandatory arguments
+
+						(*defines)[val.arguments[i]] = strings_[token.stack_pos + i + 1];
+					} else {
+						// These should be optional argument overrides
+
+						std::string str = strings_[token.stack_pos + i + 1];
+						size_t equals_pos = str.find_first_of("=");
+
+						if (equals_pos != std::string::npos) {
+							size_t argname_pos = str.substr(0, equals_pos).find_last_of(" \n") + 1;
+
+							std::string argname = str.substr(argname_pos, equals_pos - argname_pos);
+
+							(*defines)[argname] = str.substr(equals_pos + 1);
+
+							optional_arg_num++;
+
+							DBG_PREPROC << "Found override for " << argname << " in call to macro " << symbol << "\n";
+						}
+					}
+				}
+
+				// If the macro definition has any optional arguments, insert their defaults
+				if (val.optional_arguments.size() > 0) {
+					for (const auto &arg : val.optional_arguments) {
+						if (defines->find(arg.first) == defines->end()) {
+							std::ostringstream res;
+							preprocessor_streambuf *buf = new preprocessor_streambuf(target_);
```

Comment:
Why do you use pointers here istead of just a local preprocessor_streambuf  object?